torch/utils/_pytree.py

```diff
@@ -26,194 +28,124 @@
     return list(d.values()), list(d.keys())
 
 def _dict_unflatten(values: List[Any], context: Context) -> Dict[Any, Any]:
     return dict(zip(context, values))
 
-def _dict_to_str(spec: "TreeSpec", child_strings: List[str]) -> str:
-    assert spec.type == dict
-    context_child_strings = []
-    for key, child_string in zip(spec.context, child_strings):
-        context_child_strings.append(f"{key}:{child_string}")
-    return f"D({','.join(context_child_strings)})"
-
-def _maybe_str_to_dict(str_spec: str) -> Optional[Tuple[Any, Context, str]]:
-    if not str_spec.startswith("D"):
-        return None
-    context_strings, child_strings = _str_to_dict(str_spec)
-    return dict, context_strings, child_strings
-
 def _list_flatten(d: List[Any]) -> Tuple[List[Any], Context]:
     return d, None
 
 def _list_unflatten(values: List[Any], context: Context) -> List[Any]:
     return list(values)
 
-def _list_to_str(spec: "TreeSpec", child_strings: List[str]) -> str:
-    assert spec.type == list
-    return f"L({','.join(child_strings)})"
-
-def _maybe_str_to_list(str_spec: str) -> Optional[Tuple[Any, Context, str]]:
-    if not str_spec.startswith("L"):
-        return None
-    assert str_spec[1] == "("
-    assert str_spec[-1] == ")"
-    children_string = str_spec[2:-1]
-    return list, None, children_string
-
 def _tuple_flatten(d: Tuple[Any, ...]) -> Tuple[List[Any], Context]:
     return list(d), None
 
 def _tuple_unflatten(values: List[Any], context: Context) -> Tuple[Any, ...]:
     return tuple(values)
 
-def _tuple_to_str(spec: "TreeSpec", child_strings: List[str]) -> str:
-    assert spec.type == tuple
-    return f"T({','.join(child_strings)})"
-
-def _maybe_str_to_tuple(str_spec: str) -> Optional[Tuple[Any, Context, str]]:
-    if not str_spec.startswith("T"):
-        return None
-    assert str_spec[1] == "("
-    assert str_spec[-1] == ")"
-    children_string = str_spec[2:-1]
-    return tuple, None, children_string
-
 def _namedtuple_flatten(d: NamedTuple) -> Tuple[List[Any], Context]:
     return list(d), type(d)
 
 def _namedtuple_unflatten(values: List[Any], context: Context) -> NamedTuple:
     return cast(NamedTuple, context(*values))
 
-def _namedtuple_to_str(spec: "TreeSpec", child_strings: List[str]) -> str:
-    assert spec.type == namedtuple
-    context_type = {spec.context.__name__}
-    context_fields = str(spec.context._fields).replace("'", "")
-    context_type = spec.context.__name__
-    return f"N({context_type}{context_fields},{','.join(child_strings)})"
-
-def _maybe_str_to_namedtuple(str_spec: str) -> Optional[Tuple[Any, Context, str]]:
-    if not str_spec.startswith("N"):
-        return None
-    assert str_spec[1] == "("
-    assert str_spec[-1] == ")"
-    context_end_idx = str_spec.find(")") + 1
-    context_str = str_spec[2:context_end_idx]
-    children_string = str_spec[context_end_idx + 1:-1]
+def _namedtuple_serialize(context: Context) -> DumpableContext:
```

Comment:
(thinking out loud) What are the BC/FC implications of having these functions accept context instead of TreeSpec? Is TreeSpec "public"?

Reply:
Whats the BC/FC for TreeSpec? 

Lets see...if we add some value in TreeSpec, and the function accepts a context, a custom serialize function would still work since it's only dependent on the context. But I guess the function would still work if you passed in a TreeSpec? Is it possible in the future someone would want to serialize the context based on some other field in TreeSpec?

Reply:
> Whats the BC/FC for TreeSpec?

TreeSpec is public since tree_flatten (which is "public") returns it. The implementation of TreeSpec is not really supposed to be public, but it is anyways because the fields have public names (no underscore).

So, it's possible we could add additional fields to TreeSpec in the future but won't delete any existing fields.

> Is it possible in the future someone would want to serialize the context based on some other field in TreeSpec?

It is possible, but I don't think we will actually change TreeSpec anytime in the near-to-medium future. TreeSpec is pretty stable

Since we have protocols, we can always just bump the protocol version if we need to change the signature of the getstate/setstate functions, so the current design seems fine to me. E.g. in protocol==1 getstate accepts context, but in protocol==2 getstate could accept (context, some_other_field). Does that sound reasonable?

Reply:
Yeah, that makes sense!